tools/troubleshooting-sharding-ddl-locks.md

```diff
@@ -0,0 +1,152 @@
+
+If you do not need to make some DM-workers offline in the process of synchronizing sharding DDL statements, a better solution is using `stop-task` to stop the running task first, then make the DM-workers offline, and finally use `start-task` and the **new task configuration** that does not contain the already offline DM-workers to restart the task.
+
+If the owner goes offline when the owner has finished executing the DDL statement but other DM-workers have not skipped this DDL statement. For the solution, see [Condition two: a DM-worker restarts](#condition-two-a-dm-worker-restarts-or-is-unreachable-temporarily).
+
+### Manual solution
+
+1. Run `show-ddl-locks` to obtain the information of the sharding DDL lock that is currently pending synchronization. 
+
+2. Run the `unlock-ddl-lock` command to specify the information of the lock to be unlocked manually.
+
+    - If the owner of this lock is offline, you can configure the `--owner` parameter to specify another DM-worker as the new owner to execute the DDL statement.
+
+3. Run `show-ddl-locks` to check whether this lock has been successfully unlocked.
+
+### Impact
+
+After you have manually unlocked the lock, it still might exist that the lock cannot be automatically synchronized when the next sharding DDL event is received, because the offline DM-workers are included in the task configuration information.
+
+Therefore, after you have manually unlocked the DM-workers, you need to use `stop-task`/`start-task` and the updated task configuration that does not include offline DM-workers to restart the task.
+
+> **Note:** If the DM-workers that went offline become online again after you run `unlock-ddl-lock`, it means: These DM-workers will synchronize the unlocked DDL operation again. (Other DM-workers that were not offline have synchronized the DDL statement.) The DDL operation of these DM-workers will try to match the subsequent synchronized DDL statements of other DM-workers. A match error of synchronizing sharding DDL statements of different DM-workers might occur.
+
+## Condition two: a DM-worker restarts (or is unreachable temporarily)
+
+Currently, the DDL unlocking process is not atomic, during which the DM-master schedules multiple DM-workers to execute or skip the sharding DDL statement and updates the checkpoint. Therefore, it might exist that after the owner finishes executing the DDL statement, a non-owner restarts before other DM-workers skip this DDL statement. At this time, the lock information on the DM-master has been removed but the restarted DM-worker has failed to skip this DDL statement and update the checkpoint.
```

Comment:
```suggestion
Currently, the DDL unlocking process is not atomic, during which the DM-master schedules multiple DM-workers to execute or skip the sharding DDL statement and updates the checkpoint. Therefore, it might exist that after the owner finishes executing the DDL statement, a non-owner restarts before it skips this DDL statement and updates the checkpoint. At this time, the lock information on the DM-master has been removed but the restarted DM-worker has failed to skip this DDL statement and update the checkpoint.
```
a non-owner restarts before it self skips the sharding DDL statement